### ISSUE TITLE
npm version 'http://<registry-ip-address>:8080/' not found; choose one of

### ISSUE THREAD

**Issue by e-compton** [microsoft/Oryx#282]
### Bug Report

- At what date and time did you most recently experience the problem?
**2019-07-15T15:49:29**

- Where did you experience the problem? E.g. Azure Web Apps, Azure Functions, Azure Container Registry, or offline use.
**Azure Web Apps**

- If your repo is publicly available please share its URL:
n/a

- What happened?
I was trying to change our deployment to use a private NPM registry. After setting the environment `npm_config_registry` to be our private registry's address, I received an error shown below.

- What did you expect or want to happen?
I expected Oryx to install my packages from my private registry.

- How can we reproduce it?
Hopefully by setting a private repository as that environmnent variable and trying an npm install.

- Do you have log output? Please include between the backticks:
```
2019-07-15T15:49:29    benv: npm version 'http://<registry-address>:8080/' not found; choose one of:
2019-07-15T15:49:29      2 -> 2.15
2019-07-15T15:49:30      2.15 -> 2.15.9
2019-07-15T15:49:30      2.15.11
2019-07-15T15:49:30      2.15.8
2019-07-15T15:49:30      2.15.9
2019-07-15T15:49:30      3 -> 3.10
2019-07-15T15:49:30      3.10 -> 3.10.10
2019-07-15T15:49:30      3.10.10
2019-07-15T15:49:30      3.10.3
2019-07-15T15:49:30      3.9 -> 3.9.5
2019-07-15T15:49:30      3.9.5
2019-07-15T15:49:30      5 -> 5.6
2019-07-15T15:49:30      5.0 -> 5.0.3
2019-07-15T15:49:30      5.0.0
2019-07-15T15:49:30      5.0.3
2019-07-15T15:49:30      5.3 -> 5.3.0
2019-07-15T15:49:31      5.3.0
2019-07-15T15:49:31      5.4 -> 5.4.2
2019-07-15T15:49:31      5.4.2
2019-07-15T15:49:31      5.6 -> 5.6.0
2019-07-15T15:49:31      5.6.0
2019-07-15T15:49:31      6 -> 6.9
2019-07-15T15:49:31      6.9 -> 6.9.0
2019-07-15T15:49:31      6.9.0
2019-07-15T15:49:31      latest -> 6
```
```

```

**Comment by arroyc** [comment]
thanks @e-compton for reporting this, we will work on this and get back to you ..

**Comment by arroyc** [comment]
Hi @e-compton , di you get a chance to find the version you are using? or did you try recently and still have this issue?

**Comment by e-compton** [comment]
Yeah, I'm still seeing the issue on my build. The registry is a Sonatype Nexus private repository.
```
2019-07-25T13:16:29    Oryx Version       : 0.2.20190521.9, Commit: c2ee5d8bcba3b398c8390fbe2939d888dbf94352
2019-07-25T13:16:29    Build Operation ID : |S0MU6pTQPkM=.7ebb155a_
2019-07-25T13:16:30    Repository Commit  : aaef58e29432490de9df0a752b17f270904bdbf0
```

**Comment by arroyc** [comment]
ok thanks … @e-compton .. I will get back to you on this  as soon as I can repro this on my side

**Comment by e-compton** [comment]
Are there any updates on this? We're still getting it every time an Azure Web App has the `npm_config_registry` environment variable set.

**Comment by arroyc** [comment]
@e-compton  we are actively working on this, will update you once we have something. Would it be possible for you to share the app or portion of package.json for better troubleshooting from our end? I'm curious because, I wanted to see  your .npmrc  configuration and also other configuration as needed. I'm using Gemfury, but you probably using something different where your configuration may be different. We would appreciate your help if you can share some of those, that will help us to improve the product as well. Thank you again for reporting the issue and using the product

**Comment by arroyc** [comment]
@e-compton , I see http://<registry-address>:8080/ in  the log … where is your private registry hosted? Usually with app service you want a .npmrc file in your source repo which will have login/access token info of that private registry etc.

**Comment by e-compton** [comment]
So I used a PRE_BUILD_COMMAND to create the `.npmrc` just before the build started:
```
PRE_BUILD_COMMAND=echo "//$NPM_REGISTRY:_auth=$NPM_TOKEN" > .npmrc
```
The private registry is hosted on an AWS EC2 server, and all traffic goes over port 8080.

I managed to get it working by using the PRE_BUILD_COMMAND to also inject the registry:
```
echo "registry=http://$NPM_REGISTRY"$'\n'"//$NPM_REGISTRY:_auth=$NPM_TOKEN" > .npmrc
```
But if the url is set using `npm_config_registry`, then I still get that same error.

**Comment by arroyc** [comment]
I think what you are doing as a pre_build_command is a right thing. Additionally are you using npm_config_registry as an environment variable in

> application settings

 in your app service? Are you following any quickstart document or anything as such that says using npm_config_registry as environment variable ?
There are some existing issues I saw where NPM_CONFIG_REGISTRY works and npm_config_registry doesn't etc. I will try some of these options to see what works and if anything is missing and will update you accordingly. Thanks for your feedback !

**Comment by arroyc** [comment]
@e-compton , I was able to successfully deploy the app using `NPM_CONFIG_REGISTRY`. This is what my application settings look like
![image](https://user-images.githubusercontent.com/16231002/62888909-2711d400-bcf5-11e9-9850-b54de1d59209.png)
and this is what my package.json looks like
`{
  "name": "node-js-sample",
  "version": "0.2.0",
  "description": "A sample Node.js app using Express 4",
  "main": "server.js",
  "scripts": {
    "start": "node server.js"
  },
  "dependencies": {
    "print": "1.1.0",
    "express": "^4.13.3"
  },
  "engines": {
    "node": ">=4.4.7"
  },
  "keywords": [
    "node",
    "express"
  ],
  "author": "arroyc",
  "contributors": [
    "arroyc"
  ],
  "license": "MIT"
}
`
Can you please verify? Feel free to let me know if you have any questions or doubts

**Comment by arroyc** [comment]
@e-compton did you get a chance to verify this?

**Comment by e-compton** [comment]
@arroyc I'm afraid for different reasons we're now using custom docker containers to deploy our Azure Web Apps. So I no longer have access to an Oryx server to verify this on. But thanks for all your help.

**Comment by arroyc** [comment]
@e-compton , thanks for writing back to us. We are sorry seeing you moving away from Oryx, but please feel free to share any pain points/negative experiences  which made you choose custom docker containers to deploy your web apps. Your feedback will immensely help us.

**Comment by e-compton** [comment]
There were a couple of reasons why we moved away from Oryx, but I think they're more to do with Azure Web Apps than Oryx itself.

Firstly, the Oryx container doesn't rebuild until a new commit has been pushed to the repository. There is a rebuild button, but it seems to take a long time to trigger the rebuild. And there is no rebuild button when remotely deploying code through a build server like Jenkins.

Secondly, logging is very challenging to find in Azure interface. It loads really slowly and, like the rebuild button, is completely absent when remotely deploying from Jenkins. We've recently migrated from IBM Bluemix to Azure and it has taken quite a substantial amount of time just to get all the builds for our servers working.

By building the docker container on our build server we can get much faster feedback, easily redeploy, and insert our own custom logging into the container.
